unix/w0vncserver/portals/Portal.cxx

```diff
@@ -0,0 +1,252 @@
+
+  *handleToken = strdup(core::format("w0vncserver_%s", uuidStr).c_str());
+
+  free(cleanUniqueName);
+
+  vlog.debug("new request handle: %s", *requestHandle);
+  vlog.debug("new handle token: %s", *handleToken);
+}
+
+const char* Portal::newSessionHandle()
+{
+  const char* sessionHandle;
+  uuid_t uuid;
+  char uuidStr[37];
+
+  uuid_generate(uuid);
+  uuid_unparse(uuid, uuidStr);
+
+  // Valid DBus object paths may only contain
+  // the ASCII characters "[A-Z][a-z][0-9]_"
+  for (char *p = uuidStr; *p; ++p) {
+    if (*p == '-')
+      *p = '_';
+  }
+
+  sessionHandle = strdup(core::format("w0vncserver_%s", uuidStr).c_str());
+
+  vlog.debug("new session handle: %s", sessionHandle);
+
+  return sessionHandle;
+}
+
+void Portal::call(GDBusProxy* proxy, const char* method,
+                  GVariant* parameters, GDBusCallFlags flags,
+                  GDBusSignalCallback signalCallback,
+                  const char* requestHandle, void* userData)
+{
+  PortalCall* call;
+
+  call = new PortalCall();
+  call->callback = signalCallback;
+  call->userData = userData;
+  call->connection = connection_;
+  call->method = method;
+
+  if (requestHandle)
+    call->requestHandle = strdup(requestHandle);
+
+  if (signalCallback) {
+    call->signalId = g_dbus_connection_signal_subscribe(
+                      connection_,
+                      "org.freedesktop.portal.Desktop",
+                      "org.freedesktop.portal.Request",
+                      "Response",
+                      requestHandle,
+                      nullptr,
+                      G_DBUS_SIGNAL_FLAGS_NONE,
+                      [](GDBusConnection *connection,
+                         const char *sender_name,
+                         const char *object_path,
+                         const char *interface_name,
+                         const char *signal_name,
+                         GVariant *parameters_,
+                         void* user_data) {
+                           Portal* self = (Portal*)user_data;
+                           self->handleSignalResponse(connection,
+                                                      sender_name,
+                                                      object_path,
+                                                      interface_name,
+                                                      signal_name,
+                                                      parameters_,
+                                                      user_data);
+                        },
+                      call,
+                      nullptr);
+  }
+
+  g_dbus_proxy_call(proxy, method, parameters, flags, 3000, nullptr,
+                    [](GObject *source_object, GAsyncResult *res,
+				               void* data) {
+                      Portal* self = (Portal*)data;
+                      self->handleCallCb(source_object, res, data);
+                    }, userData);
+}
+
+void Portal::closeSession()
```

Comment:
This seems like it is in the wrong class? I can't see this class opening the session?